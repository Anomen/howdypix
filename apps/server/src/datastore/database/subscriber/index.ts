export * from "./updateStoreOnPhotoSave";
export * from "./updateStoreOnAlbumSave";
export * from "./updateStoreOnSearchSave";
export * from "./updateStoreOnSearchResultSave";
