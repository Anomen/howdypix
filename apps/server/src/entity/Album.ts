import { Column, ViewEntity } from "typeorm";

@ViewEntity({
  expression: `
SELECT
   P.DIR,
   P.SOURCE,
   P.PARENTDIR,
   COUNT(DISTINCT P2.ID) AS NBPHOTOS,
   COUNT(DISTINCT P3.DIR) AS NBALBUMS,
   (
      SELECT
         FILE 
      FROM
         PHOTO P4 
      WHERE
         P4.DIR = P.DIR 
         AND P4.SOURCE = P.SOURCE LIMIT 0,
         1
   )
   AS PREVIEW 
FROM
   PHOTO P 
   LEFT JOIN
      PHOTO P2 
      ON (P2.DIR = P.DIR 
      AND P2.SOURCE = P.SOURCE) 
   LEFT JOIN
      PHOTO P3 
      ON (P3.PARENTDIR = P.DIR 
      AND P3.SOURCE = P.SOURCE) 
GROUP BY
   P.DIR,
   P.SOURCE,
   P.PARENTDIR;
    `,
})
export class Album {
  @Column("text")
  dir: string;

  @Column("text")
  source: string;

  @Column("text")
  parentDir: string;

  @Column("text")
  preview: string;

  @Column("int")
  nbPhotos: number;

  @Column("int")
  nbAlbums: number;
}
