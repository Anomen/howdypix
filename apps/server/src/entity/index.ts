export * from "./Source";
export * from "./Album";
export * from "./Photo";
export * from "./Search";
export * from "./SearchResult";
