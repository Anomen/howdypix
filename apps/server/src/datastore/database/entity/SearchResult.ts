import {
  Column,
  Entity,
  PrimaryGeneratedColumn,
  ManyToOne,
  CreateDateColumn,
  UpdateDateColumn,
} from "typeorm";
import { Search } from "./Search";
import { Photo } from "./Photo";

@Entity()
export class SearchResult {
  @PrimaryGeneratedColumn()
  id: number;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column("int")
  searchId!: number;

  @Column("int")
  photoId!: number;

  @Column("int")
  public order!: number;

  @ManyToOne(() => Search, (search) => search.searchResults)
  public search!: Search;

  @ManyToOne(() => Photo, (photo) => photo.searchResults)
  public photo!: Photo;
}
