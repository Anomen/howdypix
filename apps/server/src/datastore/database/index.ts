import * as entities from "./entity";
import * as subscribers from "./subscriber";

export { entities, subscribers };
