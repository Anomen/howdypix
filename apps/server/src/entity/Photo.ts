import {
  Column,
  Entity,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  Unique,
} from "typeorm";
import { SearchResult } from "./SearchResult";
import { Album } from "./Album";

export enum PHOTO_STATUS {
  NOT_PROCESSED = "not_processed",
  PROCESSED = "processed",
}

@Entity()
@Unique(["album", "file"])
export class Photo {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  status: PHOTO_STATUS;

  @Column("int")
  inode: number;

  @Column("int")
  mtime: number;

  @Column("int")
  ctime: number;

  @Column("int")
  birthtime: number;

  @Column("int")
  size: number;

  @Column("text", { nullable: true })
  make: string;

  @Column("text", { nullable: true })
  model: string;

  @Column("int", { nullable: true })
  ISO: number;

  @Column("int", { nullable: true })
  shutter: number;

  @Column("int", { nullable: true })
  processedShutter: number;

  @Column("int", { nullable: true })
  aperture: number;

  @Column("int", { nullable: true })
  processedAperture: number;

  @Column("int", { nullable: true })
  createDate: number;

  // TODO to remove
  @Column("text")
  source: string;

  // TODO to remove
  @Column("text")
  parentDir: string;

  // TODO to remove
  @Column("text")
  dir: string;

  @ManyToOne(() => Album, (album) => album.photos)
  public album!: Album;

  @Column("text")
  file: string;

  @OneToMany(() => SearchResult, (searchResult) => searchResult.photo)
  public searchResults!: SearchResult[];
}
