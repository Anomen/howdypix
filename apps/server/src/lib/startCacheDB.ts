import { Connection } from "typeorm";
import { appDebug, generateThumbnailPaths, hjoin } from "@howdypix/utils";
import { existsSync, statSync, unlinkSync } from "fs";
import { join, parse } from "path";
import { UserConfig } from "../config";
import { Photo } from "../entity/Photo";
import { Events, EventTypes } from "./eventEmitter";
import { Source } from "../entity/Source";
import { Album } from "../entity/Album";

export async function onNewDir(
  { root, hfile }: EventTypes["newDirectory"],
  event: Events,
  connection: Connection
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const albumRepository = connection.getRepository(Album);
  const sourceRepository = connection.getRepository(Source);
  const album = await albumRepository.findOne({ where: { inode: stat.ino } });

  // Check the updated_date
  if (!album) {
    const newAlbum = new Album();
    newAlbum.dir = hfile.file ?? "";
    newAlbum.parentDir = hfile.dir ?? "";
    newAlbum.source = hfile.source;
    newAlbum.inode = stat.ino.toString();

    newAlbum.sourceLk =
      (await sourceRepository.findOne({
        source: hfile.source,
      })) ?? null;

    albumRepository.save(newAlbum);
  }
}

export async function onUnlinkDir(
  { root, hfile }: EventTypes["unlinkDirectory"],
  event: Events,
  connection: Connection
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const albumRepository = connection.getRepository(Album);
  await albumRepository.delete({ inode: stat.ino.toString() });
}

export async function onNewFile(
  { root, hfile }: EventTypes["newFile"],
  event: Events,
  connection: Connection,
  userConfig: UserConfig
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const photoRepository = connection.getRepository(Photo);
  const photo = await photoRepository.findOne({ inode: stat.ino });

  // Check the updated_date
  if (photo?.mtime === stat.mtimeMs) {
    const thumbnailsExist: boolean = generateThumbnailPaths(
      userConfig.thumbnailsDir,
      hfile
    ).reduce(
      (accumulator: boolean, tn) => accumulator && existsSync(tn.path),
      true
    );

    if (thumbnailsExist) {
      return;
    }
  }

  // Act
  event.emit("processFile", { root, hfile });
}

export async function onRemoveFile(
  { hfile }: EventTypes["newFile"],
  event: Events,
  connection: Connection,
  userConfig: UserConfig
): Promise<void> {
  const photoRepository = connection.getRepository(Photo);

  // Remove from database
  await photoRepository.delete(hfile);

  // Remove thumbnails
  generateThumbnailPaths(userConfig.thumbnailsDir, hfile).forEach((tn) => {
    unlinkSync(tn.path);
  });
}

export async function onProcessedFile(
  file: EventTypes["processedFile"],
  event: Events,
  connection: Connection
): Promise<void> {
  const albumRepository = connection.getRepository(Album);
  const photoRepository = connection.getRepository(Photo);

  const photo = new Photo();
  photo.make = file.exif.make ?? "";
  photo.model = file.exif.model ?? "";
  photo.ISO = file.exif.ISO ?? 0;
  photo.shutter = file.exif.shutter ?? 0;
  photo.processedShutter = file.exif.shutter
    ? Math.round((1 / file.exif.shutter) * 10) / 10
    : 0;
  photo.aperture = file.exif.aperture ?? 0;
  photo.processedAperture = file.exif.aperture
    ? Math.round(file.exif.aperture * 10) / 10
    : 0;
  photo.createDate = file.exif.createDate ?? 0;
  photo.inode = file.stat.inode;
  photo.mtime = file.stat.mtime;
  photo.ctime = file.stat.ctime;
  photo.birthtime = file.stat.birthtime;
  photo.size = file.stat.size;
  photo.source = file.hfile.source;
  photo.parentDir = parse(file.hfile.dir ?? "").dir ?? "";
  photo.dir = file.hfile.dir ?? "";
  photo.file = file.hfile.file ?? "";

  photo.album =
    (await albumRepository.findOne({
      dir: file.hfile.dir ?? "",
      source: file.hfile.source,
      parentDir: parse(file.hfile.dir ?? "").dir ?? "",
    })) ?? null;

  await photoRepository.save(photo);
  appDebug("db")(`Saved ${hjoin(file.hfile)}.`);
}

export async function startCacheDB(
  event: Events,
  userConfig: UserConfig,
  connection: Connection
): Promise<Connection> {
  // Insert the source entries in the database
  await Promise.all(
    Object.keys(userConfig.photoDirs).map(async (source) => {
      await Source.upsert(source, userConfig.photoDirs[source]);
    })
  );

  event.on("newFile", (params) =>
    onNewFile(params, event, connection, userConfig)
  );

  event.on("removeFile", (params) =>
    onRemoveFile(params, event, connection, userConfig)
  );

  event.on("newDirectory", (params) => onNewDir(params, event, connection));

  event.on("unlinkDirectory", (params) =>
    onUnlinkDir(params, event, connection)
  );

  event.on("processedFile", (file) => onProcessedFile(file, event, connection));

  return connection;
}
