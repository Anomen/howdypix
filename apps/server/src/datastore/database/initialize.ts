import { Connection, EntitySchema, getConnectionManager } from "typeorm";
import { map } from "lodash";
import { AppStore } from "../state";
import * as subscribers from "./subscriber";
import * as entities from "./entity";
import { AppConfig } from "../../lib/config";
import { Events } from "../../lib/eventEmitter";

export const initializeDatabase = (
  dbConfig: AppConfig["database"],
  store: AppStore,
  event: Events
): Connection =>
  getConnectionManager().create({
    type: "sqlite",
    database: dbConfig.path,
    synchronize: true,
    logging: false,
    entities: map(entities, (entity) => (entity as unknown) as EntitySchema),
    subscribers: map(
      subscribers,
      (subscriber) => subscriber(event, store) as Function
    ),
  });
