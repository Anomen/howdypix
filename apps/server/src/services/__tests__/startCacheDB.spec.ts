import { appError } from "@howdypix/utils";
import { onNewDir } from "../startCacheDB";
import { Source, Album } from "../../datastore/database/entity";

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock("@howdypix/utils", (): { [name: string]: jest.Mock } => ({
  ...jest.requireActual("@howdypix/utils"),
  appError: jest.fn(),
}));

describe("onNewDir", (): void => {
  beforeEach((): void => {
    jest.resetAllMocks();

    (appError as jest.Mock).mockReturnValue((): void => {});
  });

  test("should send an error if the source does not exist", async (): Promise<
    void
  > => {
    // @ts-ignore
    Source.fetchOne = jest.fn(null);
    // @ts-ignore
    Album.insertIfDoesntExist = jest.fn();

    await onNewDir({
      hfile: { file: "file", source: "source", dir: "dir" },
      root: "/root",
    });

    expect(appError).toHaveBeenCalled();
    expect(Album.insertIfDoesntExist).not.toHaveBeenCalled();
  });

  test("should save the album", async (): Promise<void> => {
    // @ts-ignore
    Source.fetchOne = jest.fn((): object => ({ id: 1 }));
    // @ts-ignore
    Album.insertIfDoesntExist = jest.fn();

    await onNewDir({
      hfile: { file: "file", source: "source", dir: "dir" },
      root: "/root",
    });

    expect(appError).not.toHaveBeenCalled();
    expect(
      (Album.insertIfDoesntExist as jest.Mock).mock.calls[0]
    ).toMatchSnapshot();
  });
});
