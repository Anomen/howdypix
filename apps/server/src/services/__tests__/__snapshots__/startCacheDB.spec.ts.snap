// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onNewDir should save the album 1`] = `
Array [
  Object {
    "id": 1,
  },
  "dir/file",
]
`;
